README.md
# homebridge-hunterdouglas-platinum

[![NPM version](https://img.shields.io/npm/v/homebridge-hunterdouglas-platinum)](https://npmjs.org/package/homebridge-hunterdouglas-platinum)
![License](https://img.shields.io/npm/l/homebridge-hunterdouglas-platinum)
[![Downloads](https://img.shields.io/npm/dm/homebridge-hunterdouglas-platinum.svg)](https://npmjs.org/package/homebridge-hunterdouglas-platinum)

## Hunter Douglas Platinum plug-in for [Homebridge](https://github.com/nfarina/homebridge)

<img src="https://user-images.githubusercontent.com/249172/86522462-ee9a4d80-be12-11ea-8cb2-14f7385bf7ab.jpg" width="300">

## Requirements

This plugin was recently rewritten in Typescript to use the latest capabilities of Homebridge 1.x, so it requires at a minimum:

1. Homebridge >= 1.0.0
2. Node >= 10.17.0

## Installation

<!-- 2. Clone (or pull) this repository from github into the same path Homebridge lives (usually `/usr/local/lib/node_modules`). Note: the code currently on GitHub is in beta, and is newer than the latest published version of this package on `npm` -->

1. Install homebridge using: `npm install -g homebridge`
1. Install this plug-in using: `npm install -g homebridge-hunterdouglas-platinum`
1. Update your configuration file. See example `config.json` snippet below.

## Configuration

Configuration samples (edit `~/.homebridge/config.json`):

### Direct connection via IP Address

Use this when you know the local static IP address.

```json
"platforms": [
        {
            "platform": "HunterDouglasPlatinum",
            "ip_address": "192.168.0.100"
        }
    ],
```

## Optional fields:

- `"statusPollingSeconds"` time in seconds to poll for blind positions. Default is 60 seconds.

* `"setPositionDelayMsecs"` time in msecs to delay actuually setting a blind position after it has changed. This smooths things out when you are dragging slider for a given slide. Default is 2500 msecs.

* `"setPositionThrottleRateMsecs"` rate limits how often we send set position commands to the gateway. Default is 5000 msecs.

* `"createVirtualRoomBlind"` Creates a virtual blind for each room that contains all the blinds in the room. Default is true. Changing this blind will change all blinds in the room. Position is set to the average position of all blinds in the room.

* `"createDiscreteBlinds"` Creates a blind for each real blind. Default is true. You might want to turn this off if you want only the virtual room blind.

* `"prefixRoomNameToBlindName"` Prefix the room name to each discrete blind name. Default is true.

* `"visibleBlindNames"` Comma-seperated list of blind names (prefixed with room name) to make visible. All other blinds are ignored. Only used when `createDiscreteBlinds` is true.

* `"topDownBottomUpBehavior"` Choose the desired behavior for Top-Down / Bottom-Up blinds. If `"topDown"` is specified, the slider will control the middle rail, and the blind will open downward. If `"bottomUp"` is specified, the slider will control the bottom rail, and the blind will open upward. Default is `"topDown"`. This setting has no effect on standard blinds (i.e., blinds with only "Bottom-Up" functionality).

## Sample config

```json
{
  "platform": "HunterDouglasPlatinum",
  "ip_address": "192.168.86.143",
  "statusPollingSeconds": 60,
  "createVirtualRoomBlind": true,
  "createDiscreteBlinds": true,
  "prefixRoomNameToBlindName": true
}
```

# Implemented HomeKit Accessory Types

## WindowCovering

- creates a _WindowCovering_ accessory for each discovered blind if `createDiscreteBlinds` is true

- creates a _WindowCovering_ accessory for each discovered room if `createVirtualRoomBlind` is true


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## V2.1.0 - 2020-7-5

### Added

- debounce all shade changes as a group, instead of by shade id

- this allows grouping all shades being set to the same position (with the same feature id) into a single set command to the bridge

- add `layout` to `config.schema.json` to better organize settings

## V2.0.5 - 2020-6-27

### Fixed

- only log poll errors if retry attempt is >0 to cut down on occasional `ECONNRESET` error messages

## V2.0.4 - 2020-6-27

### Fixed

- simplified the way `get` triggered a refresh
- lint/Typescript cleanup

## V2.0.3 - 2020-6-22

### Fixed

- restore Hoobs support by using `api.registerPlatform(PLUGIN_NAME, PLATFORM_NAME, HunterDouglasPlatform)`

## V2.0.2 - 2020-6-21

### Added

- cleaned up pending refresh logic and refresh on demand if needed, at most every 15 seconds for now

## V2.0.1 - 2020-6-21

### Fixed

- pass value to the CharacteristicSetCallback callback

## V2.0.0 - 2020-6-21 (MAJOR RELEASE)

### Added

- rewrite in Typescript to provide much cleaner/safer base
- minimum of Homebridge 1.1.1 is required
- support DynamicPlatformPlugin
- properly add/remove/update cached accessories on startup

### Fixed

- caching/persisting of accessories, so if bridge is unavailable when server starts the previous configured shades will remain

- add retry logic for obtaining the initial config from the controller if there is an error contacting it

- try to cleanup the way I was setting/getting/updating of service characteristics to be more proper

## V1.4.3 - 2020-6-19

### Fixed

- cleanup error logging so errors aren't logged in three different places

## V1.4.2 - 2020-5-17

### Fixed

- really fix issue #5 (off by one error)

## V1.4.1 - 2020-5-17

### Fixed

- Fixed issue #5, don't create a virtual room blind if there is only a single blind in the room.

- prettier formatting from previous pull request

## V1.4.0 - 2020-4-29

### Added

- `"topDownBottomUpBehavior"` Choose the desired behavior for Top-Down / Bottom-Up blinds. If `"topDown"` is specified, the slider will control the middle rail, and the blind will open downward. If `"bottomUp"` is specified, the slider will control the bottom rail, and the blind will open upward. Default is `"topDown"`. This setting has no effect on standard blinds (i.e., blinds with only "Bottom-Up" functionality).

## V1.3.0 - 2020-4-29

### Added

- `"setPositionThrottleRateMsecs"` rate limits how often we send set position commands to the gateway. Default is 5000 msecs.

## V1.2.3 - 2020-4-19

### Fixed

- sigh. need to read first line to make set work. log warning if it isn't what we expected

## V1.2.2 - 2020-4-19

### Fixed

- raise socket timeout back to from 30s

## V1.2.1 - 2020-4-19

### Fixed

- don't throw an error if the first line we read isn't `HunterDouglas Shade Controller`

- lower socket timeout from 30s to 15s

- don't add pollingInterval to backoff when retrying so we retry quicker on errors

- set max backoff to pollingInterval instead of pollingInterval\*20

## V1.2.0 - 2020-4-18

### Fixed

- make sure to call `connection.close()` in async convenience functions

- finally (hopefully) fix #1. Since we are using a shared Promise, make sure to have a `.catch` handler when also setting up the `.finally`

## V1.1.9 - 2020-4-11

### Fixed

- call `socket.destroy()` instead of `socket.end()`

- wrap `_pollForStatus` in a domain to see if unhandled error/exception is at the root cause of the crashing

## V1.1.8 - 2020-4-9

### Fixed

- replace `throw new BridgeError` with `this.emit('error', ...)` when handling unexpected hello response

- call `socket.end()` before emitting an error in `_handleError` (just in case)

### Added

- dump out read lines when debug is on

## V1.1.7 - 2020-1-19

### Fixed

- `_updateAccessories` `status` is null when there is an error, so add instanceof check

## V1.1.6 - 2020-1-17

### Fixed

- change `.once('error')` to `.on('error')`, otherwise if multiple errors are emitted the second one will crash node as an unhandled error

- don't start polling if no accessories are found

## V1.1.5 - 2020-1-14

### Fixed

- shouldn’t need to block during pending set

- set `Characteristic.PositionState` to `PositionState.STOPPED`

## V1.1.4 - 2020-1-12

### Fixed

- have `_command` use `_readUntil` which pull lines from buffer first instead of returning each buffered line via async

## V1.1.3 - 2020-1-12

### Added

- config `"visibleBlindNames"` Comma-seperated list of blind names (prefixed with room name) to make visible. All other blinds are ignored. Only used when `createDiscreteBlinds` is true.

- use the average of all blind positions in the room for the value of the virtual room
  blind

### Fixed

- don't log at info level in BlindAccessory

## V1.1.2 - 2020-1-12

### Fixed

- pass in log to Controller constructor

## V1.1.1 - 2020-1-12

### Fixed

- properly apply default values from config

## V1.1.0 - 2020-1-12

### Added

- config `"createVirtualRoomBlind"` Creates a virtual blind for each room that contains all the blinds in the room. Default is true. Changing this blind will change all blinds in the room.

- config `"createDiscreteBlinds"` Creates a blind for each real blind. Default is true. You might want to turn this off if you want only the virtual room blind.

- config `"prefixRoomNameToBlindName"` Prefix the room name to each discrete blind name. Default is true.

## v1.0.0 - 2020-1-4

### Added

- initial check in


src/platform.ts
import {
  API,
  WithUUID,
  DynamicPlatformPlugin,
  Logger,
  PlatformAccessory,
  PlatformConfig,
  Service,
  Characteristic,
  CharacteristicGetCallback,
} from 'homebridge'

import { PLATFORM_NAME, PLUGIN_NAME } from './settings'
import { ShadeAccessory, ShadeAccessoryContext } from './shadeAccessory'

import { Controller, ControllerConfig, ShadeStatus } from './controller'

import pThrottle from 'p-throttle'

type CharacteristicType = WithUUID<{ new (): Characteristic }>

interface PendingSetCommand {
  // context
  context: ShadeAccessoryContext
  // homekit position
  position: number
}

/**
 * HomebridgePlatform
 * This class is the main constructor for your plugin, this is where you should
 * parse the user config and discover/register accessories with Homebridge.
 */
export class HunterDouglasPlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic

  // this is used to track restored cached accessories
  private restoredAccessories: PlatformAccessory[] = []

  // this is used to track active shade accessories
  private shadeAccessories = new Map<string, ShadeAccessory>()

  // bridge controller to talk to shades
  private controller: Controller

  // fetched config
  private controllerConfig?: ControllerConfig

  // last polled shade status
  private shadeStatus?: ShadeStatus

  // for throttling set requests
  private _setTargetPositionThrottled: pThrottle.ThrottledFunction<[string[], string, number], void>

  // the time (from Date.now()) when we last initiated or did a refresh
  private lastRefreshTime = 0

  // used to debounce multiple sets to the same shadeId
  private debounceSetTimer?: NodeJS.Timeout

  // pending set commands that will be executed after debounce timer
  private pendingSetCommands = new Map<string, PendingSetCommand>()

  constructor(
    public readonly log: Logger,
    public readonly config: PlatformConfig,
    public readonly api: API,
  ) {
    this.log.debug('Finished initializing platform', PLATFORM_NAME)

    // do this first to make sure we have proper defaults moving forward
    this.applyConfigDefaults(config)

    this.controller = new Controller({
      log: this.log,
      ip_address: this.config.ip_address,
      port: this.config.port,
    })

    this._setTargetPositionThrottled = pThrottle(
      (shadeIds: string[], shadeFeatureId: string, nativePosition: number) => {
        //return Promise.resolve()
        return this.controller.setPosition(shadeIds, shadeFeatureId, nativePosition)
      },
      1, // LIMIT to 1 call within interval
      config.setPositionThrottleRateMsecs,
    )

    // When this event is fired it means Homebridge has restored all cached accessories from disk.
    // Dynamic Platform plugins should only register new accessories after this event was fired,
    // in order to ensure they weren't added to homebridge already. This event can also be used
    // to start discovery of new accessories.
    this.api.on('didFinishLaunching', () => {
      log.debug('Executed didFinishLaunching callback')
      // run the method to discover / register your devices as accessories
      this.discoverDevices(0)
    })
  }

  private discoverDevices(retryAttempt: number) {
    // test
    this.log.info('discoverDevices')

    const pollingInterval = 60 // TODO: get from config

    this.controller
      .getConfig()
      .then(config => {
        this.controllerConfig = config
        this.log.debug('got shade config', this.controllerConfig)
        this.log.info(
          `discoverDevices connected: ${this.controllerConfig.deviceId} ${this.controllerConfig.softwareVersion}`,
        )
        this.setupDiscoveredAccessories(this.controllerConfig)
      })
      .catch(err => {
        // on error, start another timeout with backoff
        const timeout = this.backoff(retryAttempt, pollingInterval)
        this.log.error(
          `discoverDevices retryAttempt: ${retryAttempt} timeout: ${timeout} error: ${err}`,
        )
        setTimeout(() => this.discoverDevices(retryAttempt + 1), timeout * 1000)
      })
  }

  private setupDiscoveredAccessories(controllerConfig: ControllerConfig) {
    // discrete shades first
    if (this.config.createDiscreteBlinds) {
      const prefixName = this.config.prefixRoomNameToBlindName

      // show only visible shades
      const visibleNames = this.config.visibleBlindNames || ''
      const visibleShades = new Set(
        visibleNames ? visibleNames.split(',').map(item => item.trim()) : [],
      )

      for (const [shadeId, shade] of controllerConfig.shades) {
        const room = controllerConfig.rooms.get(shade.roomId)
        if (room === undefined) {
          continue
        }

        if (visibleNames && !visibleShades.has((room.name + ' ' + shade.name).trim())) {
          continue
        }

        this.shadeAccessories.set(
          shadeId,
          this.configureShadeAccessory({
            displayName: prefixName ? room.name + ' ' + shade.name : shade.name,
            shadeId: shadeId,
            roomId: room.id,
            shadeFeatureId: this.getShadeFeatureId(room.shadeType),
          }),
        )
      }
    }

    // virtual room shades
    if (this.config.createVirtualRoomBlind) {
      for (const [roomId, room] of controllerConfig.rooms) {
        // only create virtual room shades if more than one shade in the room
        if (room.shadeIds.length > 1) {
          const shadeId = room.shadeIds.sort().join(',')
          this.shadeAccessories.set(
            shadeId,
            this.configureShadeAccessory({
              displayName: room.name,
              shadeId: shadeId,
              roomId: roomId,
              shadeFeatureId: this.getShadeFeatureId(room.shadeType),
            }),
          )
        }
      }
    }

    // unregister orphaned accessories
    const activeIds = Array.from(this.shadeAccessories.values()).map(accessory => accessory.UUID)
    const staleAccessories = this.restoredAccessories.filter(
      accessory => !activeIds.includes(accessory.UUID),
    )

    if (staleAccessories.length) {
      const staleNames = staleAccessories.map(accessory => accessory.displayName)
      this.log.info('unregistering accessories', staleNames)
      this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, staleAccessories)
    }

    // if we found any accessories,  start polling for status
    if (this.shadeAccessories.size) {
      this.pollForStatus(0)
    }
  }

  /**
   * This function is invoked when homebridge restores cached accessories from disk at startup.
   * It should be used to setup event handlers for characteristics and update respective values.
   */
  public configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName)

    // add the restored accessory to the accessories cache so we can track if it has already been registered
    this.restoredAccessories.push(accessory)
  }

  private configureShadeAccessory(context: ShadeAccessoryContext): ShadeAccessory {
    // generate a unique id for this shade based on context
    const uuid = ShadeAccessory.generateUUID(this, context)

    // see if an accessory with the same uuid has already been registered and restored from
    // the cached devices we stored in the `configureAccessory` method above
    const existingAccessory = this.restoredAccessories.find(accessory => accessory.UUID === uuid)

    if (existingAccessory) {
      // the accessory already exists
      this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName)

      // update the context if it has changed
      if (!ShadeAccessory.sameContext(context, <ShadeAccessoryContext>existingAccessory.context)) {
        existingAccessory.context = context
        this.log.info('Updating existing accessory:', context.displayName)
        this.api.updatePlatformAccessories([existingAccessory])
      }
      // create the accessory handler for the restored accessory
      // this is imported from `platformAccessory.ts`
      return new ShadeAccessory(this, existingAccessory)
    } else {
      // the accessory does not yet exist, so we need to create it
      this.log.info('Adding new accessory:', context.displayName)

      // create a new accessory
      const accessory = new this.api.platformAccessory(context.displayName, uuid)

      // store a copy of the device object in the `accessory.context`
      accessory.context = context

      // create the accessory handler for the newly create accessory
      const shadeAccessory = new ShadeAccessory(this, accessory)

      // link the accessory to platform
      this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory])

      return shadeAccessory
    }
  }

  /** start polling process with truncated exponential backoff: https://cloud.google.com/storage/docs/exponential-backoff */
  private pollForStatus(retryAttempt: number) {
    const pollingInterval = this.config.statusPollingSeconds

    this.refreshStatus()
      .then(() => {
        // on success, start another timeout at normal pollingInterval
        this.log.debug('_pollForStatus success, retryAttempt:', retryAttempt)
        setTimeout(() => this.pollForStatus(0), pollingInterval * 1000)
      })
      .catch(err => {
        // on error, start another timeout with backoff
        const timeout = this.backoff(retryAttempt, pollingInterval)
        if (retryAttempt > 0) {
          this.log.error('_pollForStatus retryAttempt:', retryAttempt, 'timeout:', timeout, err)
        }
        setTimeout(() => this.pollForStatus(retryAttempt + 1), timeout * 1000)
      })
  }

  /** gets status,  updates accessories, and resolves */
  private async refreshStatus() {
    try {
      this.lastRefreshTime = Date.now()
      this.shadeStatus = await this.controller.getStatus()
      this.log.debug('connected:', this.controllerConfig?.deviceId, '(getStatus)')
      // update all values
      for (const [shadeId, accessory] of this.shadeAccessories) {
        const position = this.computeShadeCurrentHomeKitPosition(shadeId)
        if (position !== undefined) {
          accessory.updateCurrentPosition(position)
          accessory.updateTargetPosition(position)
        }
      }
      return null
    } catch (err) {
      this.shadeStatus = undefined
      throw err
    }
  }

  private computeShadeCurrentHomeKitPosition(shadeId: string): number | undefined {
    if (this.shadeStatus === undefined) {
      this.log.warn('getShadeCurrentHomeKitPosition no shade status found')
      return undefined
    }
    const shadeIds = shadeId.split(',')
    const sum = shadeIds
      .map(id => this.shadeStatus?.shadeState.get(id) ?? 0)
      .reduce((sum, value) => sum + value, 0)
    return this.toHomeKitPosition(sum / shadeIds.length)
  }

  /** refresh if cached values are "stale" */
  private refreshIfNeeded(): boolean {
    const now = Date.now()
    if (now - this.lastRefreshTime > 15 * 1000) {
      // set now, so we don't trigger multiple...
      this.lastRefreshTime = now
      this.refreshStatus().catch(err => {
        this.log.error('refreshIfNeeded', err)
      })
      return true
    } else {
      return false
    }
  }

  /** convenience function to add an `on('get')` handler which refreshes accessory values if needed  */
  public triggersRefreshIfNeded(service: Service, type: CharacteristicType) {
    const characteristic = service.getCharacteristic(type)
    characteristic.on('get', (callback: CharacteristicGetCallback) => {
      // just return current cached value, and refresh if needed
      callback(null, characteristic.value)
      if (this.refreshIfNeeded()) {
        this.log.debug('triggered refresh on get', service.displayName, characteristic.displayName)
      }
    })
  }

  public setTargetPosition(context: ShadeAccessoryContext, position: number) {
    this.log.debug('setTargetPosition ', context.shadeId, context.shadeFeatureId, position)
    if (this.debounceSetTimer) {
      clearTimeout(this.debounceSetTimer)
    }

    this.debounceSetTimer = setTimeout(async () => {
      try {
        this.debounceSetTimer = undefined

        // group together all the contexts by same feature and position
        const pendingSetCommandsList = Array.from(this.pendingSetCommands.values())
        const groupedContexts = pendingSetCommandsList.reduce((result, pendingCommand) => {
          const key = pendingCommand.context.shadeFeatureId + ':' + pendingCommand.position
          if (!result.has(key)) {
            result.set(key, new Array<ShadeAccessoryContext>())
          }
          result.get(key)?.push(pendingCommand.context)
          return result
        }, new Map<string, ShadeAccessoryContext[]>())

        // now go through and send a throttled set position command for all blinds
        // with the same position and feature
        for (const [featureIdAndPos, contexts] of groupedContexts) {
          const [featureId, positionStr] = featureIdAndPos.split(':')
          const nativePosition = this.toNativePosition(parseInt(positionStr))
          const allIds = contexts.map(context => context.shadeId).join(',')
          const ids = [...new Set(allIds.split(','))]
          this.log.debug('debounced setTargetPosition:', ids, featureId, position)
          await this._setTargetPositionThrottled(ids, context.shadeFeatureId, nativePosition)
          // trigger refresh after setting. call _refreshStatus
          // instead of _refreshAccessories so we definitely fetch fresh values
          await this.refreshStatus()
        }
        this.pendingSetCommands.clear()
      } catch (err) {
        this.log.error('unable to set blind position', err)
      }
    }, this.config.setPositionDelayMsecs)
    this.pendingSetCommands.set(context.shadeId, { context: context, position: position })
  }

  private getShadeFeatureId(shadeTypeId: string) {
    // Default feature is bottom-up - Feature ID: "04"
    let shadeFeatureId = '04'
    // Special handling for top-down-bottom-up shades, which can be detected from the room's shadeTypeId
    // For top-down-bottom-up shades (shadeTypeId 02 or 13), use the top-down feature - Feature ID: "18"
    if (shadeTypeId === '02' || shadeTypeId === '13') {
      if (this.config.topDownBottomUpBehavior === 'topDown') {
        shadeFeatureId = '18'
      }
    }
    return shadeFeatureId
  }

  public generateUUID(accessorySalt: string) {
    if (this.controllerConfig !== undefined) {
      return this.api.hap.uuid.generate(this.controllerConfig.deviceId + ':' + accessorySalt)
    } else {
      this.log.error('controllerConfig is undefined')
      return ''
    }
  }

  /**
   * convert native shade position (0-255) to HomeKit (0-100)
   */
  private toHomeKitPosition(pos: number): number {
    return Math.round((pos / 255) * 100)
  }

  /**
   * convert HomeKit (0-100) to native shade position (0-255)
   *
   * @param {number} pos
   * @memberof HunterDouglasPlatinumPlatform
   */
  private toNativePosition(pos: number): number {
    return Math.round((pos / 100) * 255)
  }

  private backoff(retryAttempt: number, maxTime: number): number {
    retryAttempt = Math.max(retryAttempt, 1)
    return Math.min(Math.pow(retryAttempt - 1, 2) + Math.random(), maxTime)
  }

  public accessoryInfo(): {
    manufacturer: string
    model: string
    serialNumber: string
  } {
    if (this.controllerConfig) {
      return {
        manufacturer: 'HunterDouglas',
        // store software version in model, since it doesn't follow
        // proper n.n.n format Apple requires and model is a string
        model: this.controllerConfig.softwareVersion,
        serialNumber: this.controllerConfig.deviceId,
      }
    } else {
      this.log.error('controllerConfig is null getting accessoryInfo')
      return {
        manufacturer: 'unknown',
        model: 'unknown',
        serialNumber: '',
      }
    }
  }

  private applyConfigDefaults(config: PlatformConfig) {
    // apply defaults
    config.statusPollingSeconds = config.statusPollingSeconds ?? 60
    config.setPositionDelayMsecs = config.setPositionDelayMsecs ?? 2500
    config.setPositionThrottleRateMsecs = config.setPositionThrottleRateMsecs ?? 5000
    config.createVirtualRoomBlind = config.createVirtualRoomBlind ?? true
    config.createDiscreteBlinds = config.createDiscreteBlinds ?? true
    config.prefixRoomNameToBlindName = config.prefixRoomNameToBlindName ?? true
    config.topDownBottomUpBehavior = config.topDownBottomUpBehavior ?? 'topDown'
    config.port = config.port ?? Controller.DEFAULT_PORT
    this.log.debug('config', this.config)
  }
}
